### PR TITLE
Add main modules to ex_plasma

### PR DESCRIPTION
**Pull request by achiurizo**
## What changed and why?

This adds a few UX improvements to the client library as well as exposes main functions in the main module as our module entrypoint.

#### Transaction.new/1 [returns a result tuple](https://github.com/omisego/ex_plasma/compare/achiurizo/ux-improvements?expand=1#diff-d27add66aae6b59c6c920ead15995cdfR102)

This standardizes how we return the result with the validations. We do this currently with `Utxo.new/1` so we do it the same here as well:

```elixir
{:ok, _transaction} = Transaction.new(whatever)
{:error, {:field, :error}} = Transaction.new(bad_transaction_data)
```

#### Utxo can [build from a map](https://github.com/omisego/ex_plasma/compare/achiurizo/ux-improvements?expand=1#diff-2a228fed64e291f520cc9caac75cfbbfR128)

```elixir
{:ok, _input_utxo} = Utxo.new(%{blknum: 0, txindex: 0, oindex: 0})
{:ok, _output_utxo} = Utxo.new(%{owner: <<1::160>>, currency: <<0::160>>, amount: 1})
```

#### Add [Main functions](https://github.com/omisego/ex_plasma/compare/achiurizo/ux-improvements?expand=1#diff-eb51b4076148c3bb0129008cf2d345d9R) to ExPlasma module

```elixir
# Encoding to RLP bytes from a transaction, transaction_rlp, signed transactions
ExPlasma.encode(transaction)
# Decoding an encoded transaction byte, utxo position, transaction rlp data list, output utxo data list.
ExPlasma.decode(wahtever)
# Hash a transaction, or an encoded transaction bytes
ExPlasma.hash(wahtever)
```

#### Move Transaction types under [ExPlasma.Transaction]()

This moves all the transaction types under `ExPlasma.Transaction` instead of `ExPlasma.Transactions`.

## How was this tested?

```sh
mix test
mix test --only integration
mix test --only conformance
mix dialyzer
# run CI
```

### CODE REVIEW

File: lib/ex_plasma.ex [hunk 1]
```diff
@@ -2,4 +2,139 @@ defmodule ExPlasma do
...
+      <<217, 128, 192, 192, 128, 148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0>>
+
+       Encode a signed transaction
+      iex> {:ok, txn} = ExPlasma.Transaction.Payment.new(%{inputs: [], outputs: [], sigs: []})
+      iex> ExPlasma.encode(txn)
+      <<229, 1, 192, 192, 128, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
+  """
+  def encode(txn) when is_map(txn), do: Transaction.encode(txn)
```
Comment (InoMurko) [inline lib/ex_plasma.ex]: Dialyzer specs in this module?

File: README.md [hunk 2]
```diff
@@ -98,163 +98,6 @@ mix test --only integration
...
-})
-```
-#### Processing a standard exit
-#### Challenge a standard exit
-
-### InFlight Exits
-#### Starting an in-flight exit
-#### Processing an in-flight exit
-#### Challenge an in-flight exit
-
```
Comment (achiurizo) [inline README.md]: Removing this section of the `README` as this information is available(and more up to date) with the inline hex docs.

File: lib/ex_plasma.ex [hunk 3]
```diff
@@ -2,4 +2,143 @@ defmodule ExPlasma do
...
+      <<217, 128, 192, 192, 128, 148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0>>
+
+       Encode a signed transaction
+      iex> {:ok, txn} = ExPlasma.Transaction.Payment.new(%{inputs: [], outputs: [], sigs: []})
+      iex> ExPlasma.encode(txn)
+      <<229, 1, 192, 192, 128, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
+  """
+  @spec encode(Transaction.t() | list()) :: <<_::632>>
```
Comment (InoMurko) [inline lib/ex_plasma.ex]: nonempty_list right?

File: lib/ex_plasma.ex [hunk 4]
```diff
@@ -2,4 +2,143 @@ defmodule ExPlasma do
...
+         amount: 1,
+         blknum: nil,
+         currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
+         oindex: nil,
+         output_type: 1,
+         owner: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1>>,
+         txindex: nil
+       }}
+  """
+  @spec decode(list() | non_neg_integer() | binary()) ::
```
Comment (InoMurko) [inline lib/ex_plasma.ex]: nonempty list

File: lib/ex_plasma/client.ex [hunk 5]
```diff
@@ -38,7 +37,7 @@ defmodule ExPlasma.Client do
     |> Deposit.new()
     |> Client.deposit()
   """
-  @spec deposit(Deposit.t() | binary(), map() | list()) :: tuple()
+  @spec deposit(Transaction.t() | binary(), map() | list()) :: tuple()
   def deposit(tx_bytes, options \\ %{})
 
   def deposit(tx_bytes, options) when is_list(options),
```
Comment (InoMurko) [inline lib/ex_plasma/client.ex]: why allow options as lists?

Comment (achiurizo) [reply lib/ex_plasma/client.ex]: Removed. was playing around with a nicer API via keyword list, but not really an important need 😄

**Pull request merged**